operator/src/main/java/oracle/kubernetes/operator/steps/ManagedServerUpIteratorStep.java

```diff
@@ -168,21 +171,23 @@ public NextAction apply(Packet packet) {
 
       if (startDetailsQueue.isEmpty()) {
         return doNext(new ManagedServerUpAfterStep(getNext()), packet);
-      } else if (hasServerAvailableToStart(packet.getSpi(DomainPresenceInfo.class))) {
+      } else if (hasServerAvailableToStart(packet)) {
         numStarted.getAndIncrement();
         return doForkJoin(this, packet, Collections.singletonList(startDetailsQueue.poll()));
       } else {
         return doDelay(this, packet, SCHEDULING_DETECTION_DELAY, TimeUnit.MILLISECONDS);
       }
     }
 
-    private boolean hasServerAvailableToStart(DomainPresenceInfo info) {
-      return ((numStarted.get() < info.getNumScheduledServers(clusterName))
-              && (canStartConcurrently(info.getNumReadyServers(clusterName))));
+    private boolean hasServerAvailableToStart(Packet packet) {
+      DomainPresenceInfo info = packet.getSpi(DomainPresenceInfo.class);
+      String adminServerName = ((WlsDomainConfig) packet.get(DOMAIN_TOPOLOGY)).getAdminServerName();
+      return ((numStarted.get() <= info.getNumScheduledManagedServers(clusterName, adminServerName)
+              && (canStartConcurrently(info.getNumReadyManagedServers(clusterName, adminServerName)))));
     }
 
     private boolean canStartConcurrently(long numReady) {
-      return ((this.maxConcurrency > 0) && (numStarted.get() < (this.maxConcurrency + numReady - 1)))
+      return ((this.maxConcurrency > 0) && (numStarted.get() < (this.maxConcurrency + numReady)))
```

Comment:
This would be a good time to do some refactoring. The expression is not so easy to read and understand.

I would start by extracting maxConcurrency == 0 into a method named 'ignoreConcurrencyLimits' and list that first in the expression, not last. Then you don't really need `this.maxConcurrency > 0` as that is implied (unless there is some support for negative values). 

The remaining part is still too complex and not obvious. So extract `numStarted.get()` into its own method, `getNumServersStarted` as there is no reason for all of the callers to know that there is an Atomic backing the value. Then can you explain what the rest of the logic is?

Reply:
Made the refactoring changes. The remaining logic is to check if the number of started servers is less than (maxConcurrency + ready managed servers for cluster "clusterName"). So if maxConcurrency is `1` and one server is started, it'll wait for that server to come to ready state before starting a new server. Please see https://github.com/oracle/weblogic-kubernetes-operator/pull/1855#issuecomment-671665520 for the pseudo code of this logic.

Reply:
I looked at the comment. The pseudo code doesn't explain the logic, either. 

It appears to me that the intent of `numStarted < maxConcurrency + numReady` is to look at the number of pods currently starting up (I think K8S calls that `waiting`) and compare that how many may be started at once. If so, the actual comparison should be:

`(numStarted - numReady) < maxConcurrency` or

`numWaiting < maxConcurrency`.

If so, I would extract a `numPodsWaiting` method:

```
// returns the number of pods currently starting up
private int numWaiting(int numReady) {
  return getNumStarted() - numReady;
}
```

and compare that to the maxConcurrency.

Reply:
Yes, that's correct. I think k8s calls pods that are currently starting up as `not ready` or `inactive`. We can go with `waiting` if that clarifies the intention of code. I'll go ahead and push above change.